### ISSUE TITLE
Issues with June 5, 2021 Tutorial

### ISSUE THREAD

**Issue by neellab-umd** [thibautjombart/adegenet#307]
Hi There Zhian -

I am really grateful that you updated the adegenet tutorial.  It has been a chore to learn how to use it while wading through deprecated functions.  I do not envy you trying to keep it up.

I know you tried to get it posted really quickly so you probably didn't notice the places where the code fails.  Hopefully pointing these out is helpful to you (and others).

First, you may no longer have wordcloud installed because the library() call to that package fails (page 61).

On page 66 you get this error message

```r
## Error in xj[i, , drop = FALSE]: incorrect number of dimensions
```

That one interests me because I am getting it today on code I know use to work in which I am converting genind objects to genclone objects.  I can't see anything I am doing so I am thinking something behind the scenes got updated and matrices are not stored in the same way or place.

The failure above generates follow-on errors in the next three code blocks (page 66 and 67) because the intended object is not created.

On page 71 pairwise.fst() is not found.

Again, I mean to be helpful in pointing these out.  I truly appreciate all you do.

**Comment by zkamvar** [comment]
Thank you for reporting this! I'm looking into it and I can confirm that pairwise.fst() not found is 100% my fault for not seeing it. 

The other error is _very_ curious because it's coming from the `lm()` call. Both of the objects are distance matrices and it's failing when attempting to call `model.frame()` for the formula. I tried to create a reproducible example without adegenet loaded, but couldn't get the error until I loaded adegenet, which means that something very unexpected is happening (I am currently digging into this)

``` r
a <- dist(1:3)
b <- dist(6:4)
model.frame(a ~ b) # works
#>   a b
#> 1 1 1
#> 2 2 2
#> 3 1 1
library(adegenet)
#> Loading required package: ade4
#> Registered S3 method overwritten by 'spdep':
#>   method   from
#>   plot.mst ape
#> 
#>    /// adegenet 2.1.3 is loaded ////////////
#> 
#>    > overview: '?adegenet'
#>    > tutorials/doc/questions: 'adegenetWeb()' 
#>    > bug reports/feature requests: adegenetIssues()
model.frame(a ~ b) # fails
#> Error in xj[i, , drop = FALSE]: incorrect number of dimensions
```

<sup>Created on 2021-06-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by zkamvar** [comment]
The error is coming from [these lines in the R source:](https://github.com/wch/r-source/blob/3990b64906a2ebdaf8690f116226ed3d2eec82c0/src/library/base/R/dataframe.R#L720-L724)

```r
    for(j in seq_along(x)) {
        xj <- xx[[ sxx[j] ]]
        ## had drop = drop prior to 1.8.0
        x[[j]] <- if(length(dim(xj)) != 2L) xj[i] else xj[i, , drop = FALSE]
    }
```
The reason it's failing is because `xj` is a single column of the data frame, but it has retained its "dist" class, which means that it has two dimensions, so the last line of this snippet says, "subset this two-dimensional column by the first elements." The problem is that even though "dist" objects have multiple dimensions reflecting the legs of the triangle, they do not subset like matrices. 

Something about adegenet being loaded is changing the behavior of dist objects.

**Comment by zkamvar** [comment]
I don't know how... I don't know why... but somehow the `dim()` primitive is now returning dimensions for dist objects after adegenet is being loaded :(

``` r
dat <- structure(
  list(
    a = structure(c(1, 2, 1), Size = 3L, Diag = FALSE, Upper = FALSE, method = "euclidean", call = dist(x = 1:3), class = "dist"), 
    b = structure(c(1, 2, 1), Size = 3L, Diag = FALSE, Upper = FALSE, method = "euclidean", call = dist(x = 6:4), class = "dist") 
  ),
  class = "data.frame", 
  row.names = c(NA,  3L),  
  terms = a ~ b
)
dim(dat[[1]])
#> NULL
library(adegenet)
#> Loading required package: ade4
#> Registered S3 method overwritten by 'spdep':
#>   method   from
#>   plot.mst ape
#> 
#>    /// adegenet 2.1.3 is loaded ////////////
#> 
#>    > overview: '?adegenet'
#>    > tutorials/doc/questions: 'adegenetWeb()' 
#>    > bug reports/feature requests: adegenetIssues()
dim(dat[[1]])
#> [1] 3 3
```

<sup>Created on 2021-06-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

This is what's causing the error.

**Comment by zkamvar** [comment]
Until I can find out what is going on, the solution for this is to convert the matrices to vectors:

```r
lm(as.vector(Dgen) ~ as.vector(Dgeno))
```

**Comment by zkamvar** [comment]
Okay, so the reason this is only suddenly behaving weird:

1. adegenet depends on sf which depends on e1071
2. e1071 recently picked up proxy as a dependency
3. proxy adds a "dim" method for dist objects

**Comment by neellab-umd** [comment]
This is some seriously impressive detective work. It looks like it took up
your Saturday.  I am assuming this change is the cause of one of my other
script breaking.

The part of my script does not use adegenet directly, rather it is using
the package diveRsity - I am figuring it either calls adegenet in the
background or also relies on sf or e1071.  I use the divBasic function on a
genepop format file I created from a genind object. It has always worked
and it is now seen as having the wrong dimensions.  I could not see what
had changed.  I don't see how to easily fix the problem because I don't
know where within divBasic a matrix is being called.

Is this something that one can ask the proxy package keepers to address?

Maile

**Comment by zkamvar** [comment]
The maintainer of proxy and the maintainer on e1071 are the same person, so I will email them and alert them to the issue. 

You might also be able to [install a specific version](https://support.rstudio.com/hc/en-us/articles/219949047-Installing-older-versions-of-packages) of e1071 to avoid the proxy issue (note: it needs a c compiler):

```r
remotes::install_version("e1071", "1.7-4")
```

I'm _fairly_ sure that would work. Please report back if it does.

**Comment by neellab-umd** [comment]
I installed version 1.7-4 of e1071 as you suggest and got rid of the newer
one.  Unfortunately I still get this error:

Error in x[, 1] : incorrect number of dimensions.

I narrowed the problem down to something in the part of the divBasic
function that implements bootstrapping.  If I leave out bootstrapping from
my function call, divBasic runs fine.  That lets me get the basic
statistics but not confidence intervals.

I know that divBasic has nothing to do with packages you maintain and I
know I have not given you reproducible code so I appreciate your help.

Maile
